### PR TITLE
provide "loc : Loc.t" binding within "VERNAC COMMAND EXTEND" rules

### PR DESCRIPTION
**Pull request by ghost**
Originally, it was not possible to define a new vernacular command
in the following way:
```
VERNAC COMMAND EXTEND Cmd6 CLASSIFIED AS QUERY
    [ "SomeCmd" ] -> [ Feedback.msg_notice ?loc (Pp.str "some message") ]
END
```
because `loc : Loc.t` was not bound.

This commit fixes that, i.e. the location of the custom Vernacular command
(where it occurs in the `*.v` file) is made available as `loc` variable bound on the right side
of `->` .

### CODE REVIEW

**Comment by matejkosik** [comment]
@herbelin We talked about this in Le Croisic.

**Review by herbelin — APPROVED** [review]
@matejkosik: And, indeed, nothing to add from my point of view. Ready to merge.

**Comment by matejkosik** [comment]
@Zimmi48 AppVeyor reports:
```
Build execution time has reached the maximum allowed time for your plan (60 minutes).
```
Is this something that need to be addressed?

**Comment by Zimmi48** [comment]
@matejkosik I'm not knowledgeable on the Appveyor infrastructure, but my understanding is that right now this is crap, not working at all (failing most of the times) and thus its results are simply ignored. That's a pity and I hope it can be fixed soon but I won't be the one spending time trying to understand it.

**Review by ppedrot — APPROVED** [review]
Indeed, that's a valuable change.

**Comment by matejkosik** [comment]
@Zimmi48 Ok, I've squashed the two commits into one (and removed the `needs: squashing` label.

File: API/API.mli [hunk 1]
```diff
@@ -5666,7 +5666,7 @@ module Vernacinterp :
 sig
   type deprecation = bool
 
-  type vernac_command = Genarg.raw_generic_argument list -> unit -> unit
+  type vernac_command = Genarg.raw_generic_argument list -> Loc.t option -> unit
```
Comment (ejgallego) [inline API/API.mli]: Maybe this should be `Genarg.raw_generic_argument list Loc.located` ?

Comment (herbelin) [inline API/API.mli]: In the current model (for better error reporting as far as I understood), the evaluation is done in two steps and this requires a partial application (see the call to `hunk`).

Comment (matejkosik) [inline API/API.mli]: @ejgallego I am not sure if I understand why would that be beneficial.

Comment (ejgallego) [inline API/API.mli]: @herbelin IMHO this is orthogonal, the thunk could be kept in both approaches. @matejkosik the real question here to know what we should use is "is the first argument a located object"? If so, IMHO, we should then use the existing abstraction.

Note of course that not all objects involving locations do indeed meet this criteria.

Comment (matejkosik) [inline API/API.mli]: @ejgallego I let @herbelin reply to you as he knows more about these things than me.

To me, it would help if I could look at a PR that would realize what you have in mind.

Comment (herbelin) [inline API/API.mli]: @ejgallego:
> the thunk could be kept in both approaches.

After all, if needed, yes.

Comment (maximedenes) [reply API/API.mli]: @matejkosik Can you take @ejgallego's comment into account and update this PR? Thanks!

Comment (matejkosik) [reply API/API.mli]: Do I interpret the suggestion of @ejgallego correctly? Do we really want to change the types of all variables bound in `VERNAC COMMAND EXTEND` macro clauses?

(In other words, we would break all plugins which use this macro.)

Comment (maximedenes) [reply API/API.mli]: Ok, after talking offline I think there was a misunderstanding. @ejgallego suggested to move to `Genarg.raw_generic_argument list Loc.located`, not `Genarg.raw_generic_argument located list`, which was what @matejkosik didn't know how to do.

Comment (ejgallego) [reply API/API.mli]: Yeah, anyways I dunno what makes sense here, [certainly breaking the API makes no sense]. At this level we tend to have two kind of objects: objects that always carry a location, and objects that may carry a location. I would be cool to handle them with different types, we got quite a bit of space to improve the API here.

**Comment by matejkosik** [comment]
I am closing this PR because PR #952 (which documents relevant types/functions) is not mergable.

**Comment by maximedenes** [comment]
?! I don't understand the link between the two.

**Comment by herbelin** [comment]
I'm in favor of merging this PR. As for #952, one part is straighforwardly needed (adding directory `vernac` to the list of `mli`). Other parts are more subjective. I'd be happy with my propositions but I'll not fight for it and will be happy with alternative propositions. 

Otherwise, the following changes to coqdoc [here](https://github.com/matejkosik/coq/commits/master%2BEXPERIMENT%2Bfix_ocamldoc_warnings) are ok to me: the three workarounds around `S` modules, the `@raise` tag fix, the superflous `Util`. I'm unsure about the reformatting (is it needed for coqdoc?), with the expanded module signature. As for `@c`in a comment, it is correct but should be escaped since it denotes a Coq explicit application, not an ocamldoc tag.

@matejkosik: is it ok if I reopen the PR?

**Comment by matejkosik** [comment]
@herbelin Since I deleted my Coq repository, it is not longer possible to reopen the PR. But I have reforked Coq repository and re-pushed the [same branch](https://github.com/matejkosik/coq/tree/master+VERNAC_COMMAND_EXTEND+Loc.t).

To integrate it, unfortunatelly, another PR must be created. My attempt failed, but maybe you will have more luck/skills.

(I got deadlocked by trying to do some ambitious things with `ocamldoc` which was a mistake. I suspect that it is not "fiercely maintained". The final nail to the coffin (for me) was when I found out that even if it generates warnings about suspicious things (as well as bogus warnings), it does not tell you the corresponding file-name, line-numbers.)

**Comment by ejgallego** [comment]
> The final nail to the coffin (for me) was when I found out that even if it generates warnings about suspicious things (as well as bogus warnings), it does not tell you the corresponding file-name, line-numbers.)

I guess a patch would be welcome then, right?

**Comment by matejkosik** [comment]
> I guess a patch would be welcome then, right?

I see two good reasons when it would make sense to do so:
1. if `ocamldoc` solved some of the problems we have,
1. or if we were payed to maintain it.

Concerning checking breakage of links from the documentation to the actual source-code-elements --- I just use `compilerlibs`.
